File: AtCoder/Greens/python/ABC125-C-GCD_on_Blackboard.py
from math import gcd
N = int(input())
A = list(map(int, input().split()))

G1 = [A[0]] + [0] * N
G2 = [0] * N + [A[N - 1]]
for i in range(N):
    G1[i + 1] = gcd(G1[i], A[i])
    G2[N - i - 1] = gcd(G2[N - i], A[N - i - 1])
res = 0
for i in range(N):
    if i == 0:
        res = max(res, G2[i + 1])
    elif i == N - 1:
        res = max(res, G1[i])
    else:
        res = max(res, gcd(G1[i], G2[i + 1]))
print(res)

File: AtCoder/Greens/cpp/ABC125-C-GCD_on_Blackboard.cpp
#include <bits/stdc++.h>
#define MAX_N 100'000
using namespace std;
typedef long long ll;

ll N;
ll A[MAX_N];
ll G1[MAX_N + 1];
ll G2[MAX_N + 1];

int main() {
  cin >> N;
  for (ll i = 0; i < N; ++i) {
    cin >> A[i];
  }
  G1[0] = A[0];
  G2[N] = A[N - 1];
  for (ll i = 0; i < N; ++i) {
    G1[i + 1] = __gcd(G1[i], A[i]);
    G2[N - i - 1] = __gcd(G2[N - i], A[N - i - 1]);
  }
  ll res = 0;
  for (ll i = 0; i < N; ++i) {
    if (i == 0) {
      res = max(res, G2[i + 1]);
    } else if (i == N - 1) {
      res = max(res, G1[i]);
    } else {
      res = max(res, __gcd(G1[i], G2[i + 1]));
    }
  }
  printf("%lld\n", res);
}